### PR TITLE
Defer modify replication group if any of its node groups is not in available state

### PR DESCRIPTION
**Pull request by kumargauravsharma**
Issue #529 

Description of changes:
- Added code to check node groups state before invoking modify replication group
- Updated unit tests to include corresponding test scenarios

```make test``` passed.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by A-Hilaly** [comment]
I'm wondering if we should split pkg/ unit tests and service custom code unit tests. I think it makes more sense to run a specific service unit tests using something like `make test SERVICE=elasticache`

**Comment by jaypipes** [comment]
> I'm wondering if we should split pkg/ unit tests and service custom code unit tests. I think it makes more sense to run a specific service unit tests using something like `make test SERVICE=elasticache`

Maybe eventually, sure.

File: services/elasticache/pkg/resource/replication_group/custom_update_api.go [hunk 1]
```diff
@@ -35,12 +35,22 @@ func (rm *resourceManager) CustomModifyReplicationGroup(
...
+			if nodeGroup.Status == nil || *nodeGroup.Status != "available" {
+				allNodeGroupsAvailable = false
+				break
+			}
+		}
+	}
+	if latestRGStatus == nil || *latestRGStatus != "available" || !allNodeGroupsAvailable {
 		return nil, requeue.NeededAfter(
 			errors.New("Replication Group can not be modified, it is not in 'available' state."),
 			requeue.DefaultRequeueAfterDuration)
```
Comment (jaypipes) [inline services/elasticache/pkg/resource/replication_group/custom_update_api.go:51]: eventually we might want to have this interval be configurable? or at least specific to Elasticache?

File: services/elasticache/pkg/resource/replication_group/custom_update_api.go [hunk 2]
```diff
@@ -35,12 +35,22 @@ func (rm *resourceManager) CustomModifyReplicationGroup(
...
 	}
-
+
```
Comment (jaypipes) [inline services/elasticache/pkg/resource/replication_group/custom_update_api.go:53]: whitespace here?

File: services/elasticache/pkg/resource/replication_group/custom_update_api_test.go [hunk 3]
```diff
@@ -16,6 +16,8 @@ package replication_group
 import (
 	"context"
 	"fmt"
+	"github.com/aws/aws-controllers-k8s/pkg/requeue"
+	"github.com/pkg/errors"
```
Comment (jaypipes) [inline services/elasticache/pkg/resource/replication_group/custom_update_api_test.go:20]: this is only a test file of course, but common practice is to put third-party imports together separated by a newline after standard library imports.

File: services/elasticache/pkg/resource/replication_group/custom_update_api_test.go [hunk 4]
```diff
@@ -223,6 +234,69 @@ func TestCustomModifyReplicationGroup(t *testing.T) {
...
+		for _, nodeGroup := range latest.ko.Status.NodeGroups {
+			nodeGroup.Status = &unavailableStatus
+		}
+		var diffReporter ackcompare.Reporter
+		var ctx context.Context
+		res, err := rm.CustomModifyReplicationGroup(ctx, desired, latest, &diffReporter)
+		assert.Nil(res)
+		assert.Nil(err)
+	})
+}
```
Comment (jaypipes) [inline services/elasticache/pkg/resource/replication_group/custom_update_api_test.go:298]: ++ nice unit tests.

**Review by jaypipes — APPROVED** [review]
Good work @kumargauravsharma. Couple little nits inline but you can address in a followup patch whenever...

**Pull request merged**